File: core/src/ru/myitschool/cubegame/story/quest/ReachRoomQuest.java
package ru.myitschool.cubegame.story.quest;

import com.badlogic.gdx.utils.Array;
import ru.myitschool.cubegame.dungeon.Room;
import ru.myitschool.cubegame.effects.CellEffect;
import ru.myitschool.cubegame.effects.FloorEffect;
import ru.myitschool.cubegame.entities.Entity;
import ru.myitschool.cubegame.skills.Target;
import ru.myitschool.cubegame.tiles.DungeonTile;

/**
 * Created by Voyager on 08.04.2018.
 */
public class ReachRoomQuest extends Quest {

    public ReachRoomQuest(int roomPos, int deltaPos, int roomsMax) {
        super(roomPos, deltaPos, roomsMax);
    }

    @Override
    public Room modifyRoom(Room room) {
        Array<Target> targets = new Array<>();
        for (int i = 0; i < room.getWidth(); i++) {
            for (int j = 0; j < room.getHeight(); j++) {
                if (room.getCell(i, j) != null && DungeonTile.getTile(room.getCell(i, j)).isReachable()){
                    targets.add(new Target(i, j));
                }
            }
        }

        FloorEffect effect = null;
        CellEffect cellEffect = new CellEffect(effect){
            @Override
            protected void onStepTo(Entity entity) {
                super.onStepTo(entity);
                targetReached();
            }
        };
        effect = new FloorEffect(targets, cellEffect, getExitColor(), true, false);

        room.addEffect(effect);
        return room;
    }

    @Override
    public void exitRoomOpened() {

    }

    @Override
    public void targetReached() {
        complete();
    }
}

File: core/src/com/wolg_vlad/dcrawler/story/quest/ReachRoomQuest.kt
package com.wolg_vlad.dcrawler.story.quest

import com.badlogic.gdx.graphics.Texture
import com.wolg_vlad.dcrawler.dungeon.Room
import com.wolg_vlad.dcrawler.effects.CellEffect
import com.wolg_vlad.dcrawler.effects.FloorEffect
import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.entities.skills.Target
import com.wolg_vlad.dcrawler.ui.tiles.DungeonTile

/**
 * Created by Voyager on 08.04.2018.
 */
class ReachRoomQuest(roomPos: Int, deltaPos: Int, roomsMax: Int) : Quest(roomPos, deltaPos, roomsMax) {
    override fun modifyRoom(room: Room): Room {
        val targets = mutableListOf<Target>()
        for (i in 0 until room!!.width) {
            for (j in 0 until room.height) {
                if (room.getCell(i, j) != null && DungeonTile.getTile(room.getCell(i, j)!!).groundTile) {
                    targets.add(Target(i, j))
                }
            }
        }
        var effect: FloorEffect? = null
        val cellEffect: CellEffect = object : CellEffect(effect) {
            override val icon: Texture? = null
            override val name: String = ""
            override val description: String = ""
            override val skillUse: Boolean = false
            override val positive: Boolean = false
            override val stackable: Boolean = false
            override val stackSize: Int = 0
            override val expiring: Boolean = false
            override val id: String = "main.dcrawler.effect.quest.reach_room"
            override val expireTurns: Int = 0

            override fun onStepTo(entity: Entity) {
                super.onStepTo(entity)
                targetReached()
            }
        }
        effect = FloorEffect(targets, cellEffect, Quest.exitColor, true, false)
        room.addEffect(effect)
        return room
    }

    override fun exitRoomOpened() {}
    override fun targetReached() {
        complete()
    }
}